compiler/rustc_passes/src/entry.rs

```diff
@@ -164,12 +166,47 @@ fn configure_main(
         Some((tcx.hir().local_def_id(hir_id), EntryFnType::Main))
     } else if let Some((hir_id, _)) = visitor.main_fn {
         Some((tcx.hir().local_def_id(hir_id), EntryFnType::Main))
+    } else if let Some(local_def_id) = tcx.hir().krate().main_by_path {
+        if !tcx.features().main_by_path {
+            let sp = tcx
+                .hir()
+                .span_if_local(local_def_id.to_def_id())
+                .unwrap_or(tcx.hir().krate().item.inner);
+            feature_err(
+                &tcx.sess.parse_sess,
+                sym::main_by_path,
+                sp,
+                "Using a re-export of a function as entry point `main` is experimental",
+            )
+            .emit();
+            no_main_err(tcx, visitor);
+            return None;
+        }
+        if let Some(DefKind::Fn) = tcx.hir().opt_def_kind(local_def_id) {
+            Some((local_def_id, EntryFnType::Main))
+        } else {
+            let sp = tcx.hir().span_if_local(local_def_id.to_def_id());
+            invalid_main_reexport_err(tcx, sp);
+            None
+        }
     } else {
         no_main_err(tcx, visitor);
         None
     }
 }
 
+fn invalid_main_reexport_err(tcx: TyCtxt<'_>, sp: Option<Span>) {
+    // There is something called `main`, but it is not a function definition.
+    let mut err = tcx.sess.struct_span_err(
+        sp.unwrap_or(DUMMY_SP),
+        &format!(
+            "non-function item with name `main` is found in crate `{}`",
```

Comment:
This is a breaking change, if something named `main` exists in root module in value namespace, but it is not a function, then it should simply be ignored.

Reply:
Ok, converted it to a note for the `no_main` error.